rfcs/2021-04-13-7027-core-extraction.md

```diff
@@ -0,0 +1,319 @@
+
+As a sub-alternative to "Reduce Features" I will note that today we often
+encourage our engineers to compile with `--no-default-features` flagged on. This
+_does_ reduce build times and can, depending on the area of the code base you're
+working on, improve iteration loop speed. We might make this the official
+default. This alternative would require us to modify our "release" build to flip
+on the specific features we intend to ship. However, while this alternative does
+potentially reduce iteration costs and experimentation costs depending on the
+area being worked it does nothing for reducing the burden for casual
+contributors nor does it lead to more focused tests. A non-representative
+default may be surprising to casual contributors, leading to CI dings that would
+otherwise have happened locally.
+
+### Top-Level is Core, Concepts Become Packages
+
+This alternative pushes as much as possible into packages, leaving whatever
+remains as the "core" of vector. All transforms would go into a "transforms"
+package with sub-packages for large conceptual areas, similar for sinks and
+sources. VRL would remain as-is and we'd need to debate in the future what is
+and isn't "core". Once done this alternative would reduce iteration costs on
+core and, if documented, give casual contributors guideposts about where to add
+code. Experimentation costs could also be reduced, with packages in need of
+modification because of changes to core being dealt with only after the changes
+to core as shown to be worthwhile. This alternative does not make for a more
+focused testing environment. With "core" as the top-level package we can't
+follow the sub-package model of exposing internal interfaces for experimentation
+as easily: our `vector` binary lives at the top-level, what if we want to create
+a `core` binary at the top-level that takes no configuration and just sends data
+through itself for testing purposes? Is that confusing to our users? I, at
+least, would find it so. Should configuration be a "core" issue or a package
+used exclusively by "core"? How granular should a package be? Must all sources
+go in a "sources" package or can `file-source` continue, or should it be a crate
+in a package? These are important questions and they'll need answered ahead of
+time in this alternative. That is, "core" is extracted by bulk movement of code
+in this approach.
+
+### Core is Just Another Package
+
+This alternative is the inverse of the previous: core is just another
+package. We extract what code is "core" to vector and move it into its own
+package, rigged with private tests, interfaces in the manner of VRL. This will
+allow work on core to happen in a similar manner to VRL today, with tighter
+iteration and more focus. Experimentation in core specifically need not include
+the rest of the project, until after the experiment is shown to be
+positive. Casual contributors will be unlikely to work with core -- most outside
```

Comment:
Integration of such experiments will, of course, still have to include the rest of the project, unless we are able to come up with an interface layer that is somehow magically both zero cost and opaque to the underlying details.

Reply:
Yep, yep. The goal here being that we can experiment in a relatively cheap idea, toss the work if it's not effective in the limited context of core and then repeat the experiment once we update the rest of the project.